### ISSUE TITLE
login problem with long passwords

### ISSUE THREAD

**Issue by sandormarton** [roundcube/roundcubemail#6247]
Hi,

I have a weird problem, i'm unable to login with 9+ char long passwords, and is driving me crazy.
Have a fairly straightforward Postfixadmin based postfix+dovecot setup, Roundcube 1.3.6 ( but noticed on earlier versions too ) , PHP 5.6.30 .
Issue : i can't login with passwords longer than 9 characters, while smaller passwords work . Same user, freshly created. At the same time POP/IMAP with Thunderbird will work .
Error being :   
`[13-Apr-2018 01:28:50 +0300]: <dnp2k472> IMAP Error: Login failed for testsandor@testdomain.com from 81.181.202.157. AUTHENTICATE PLAIN: Authentication failed. in /var/www/rcube/program/lib/Roundcube/rcube_imap.php on line 196 (GET /email/?_task=mail&_action=list&_refresh=1&_mbox=INBOX&_remote=1&_unlock=loading1523572124563&_=1523572124518)
[13-Apr-2018 01:28:50 +0300]: <dnp2k472> IMAP Error: Login failed for testsandor@testdomain.com from 81.181.202.157. AUTHENTICATE PLAIN: Authentication failed. in /var/www/rcube/program/lib/Roundcube/rcube_imap.php on line 196 (GET /email/?_task=mail&_action=getunread&_page=1&_remote=1&_unlock=0&_=1523572124519)
`
I'm attaching 2 files, imap log from Roundcube with a short password, and when login is working ( password  : 'qwerty'; and a log with the password 'qwerty1234' and when Roundcube doesn't login .
[badlogin.txt](https://github.com/roundcube/roundcubemail/files/1905243/badlogin.txt)
[goodlogin.txt](https://github.com/roundcube/roundcubemail/files/1905244/goodlogin.txt)

Since Roundcube's imap log doesn't contain the passwords, i captured with tcpdump the conversation with dovecot, and the noticed something, Roundcube will try first time the good password, then scrambles it . In the capture i see these 4 AUTH lines : 
`AUTHENTICATE PLAIN AHRlc3RzYW5kb3JAdGVzdGRvbWFpbi5jb20AcXdlcnR5MTIzNA==
AUTHENTICATE PLAIN AHRlc3RzYW5kb3JAdGVzdGRvbWFpbi5jb20AMzRlcnR5MTIzNA==
AUTHENTICATE PLAIN AHRlc3RzYW5kb3JAdGVzdGRvbWFpbi5jb20AMzRlcnR5MTIzNA==
AUTHENTICATE PLAIN AHRlc3RzYW5kb3JAdGVzdGRvbWFpbi5jb20AMzRlcnR5MTIzNA==`
First uses the ok password, but second and the others will not.

Any idea?

**Comment by sandormarton** [comment]
Yes, i just switched to PLAIN so i can check the transmitted password.
Log attached with LOGIN 
[ll.txt](https://github.com/roundcube/roundcubemail/files/1906659/ll.txt)

**Comment by alecpl** [comment]
Ok, but `A0002 LOGIN ****** [37]` does not tell me much. Do you test with disabled all plugins. I'm using long passwords without issues.

**Comment by alecpl** [comment]
I was rather thinking about adding some debug code here and there. Like `rcube::console($var);` somewhere in program/lib/Roundcube/rcube.php (where you can find encrypt/decrypt methods). That would log `$var` variable content to the logs/console file.

**Comment by alecpl** [comment]
or encrypt() creates malformed result. There's a few variables in there that you have to debug, $ckey, $method, $iv, $iv_size. Maybe their values will give us some hints.

**Comment by sandormarton** [comment]
Patch to program/lib/Roundcube/rcube.php : 
[cons1patch.txt](https://github.com/roundcube/roundcubemail/files/1916058/cons1patch.txt)

Was getting this in console : 
[cons1.txt](https://github.com/roundcube/roundcubemail/files/1916065/cons1.txt)

Was looking at the result, and i figured i should try to change cipher. So i changed in config to AES-256-CBC, and to my surprise the password started working !
Since DES-EDE3-CBC cipher length is 8 byte, while AES-256-CBC is 16, i figured i should try with a 17 char password, and yes, that didn't work with AES-256-CBC ( while a 16 char worked. Also the 17 char worked with a mail client )
console from the 17 char test  : 
[cons2.txt](https://github.com/roundcube/roundcubemail/files/1916114/cons2.txt)

So the bug is somehow connected to cipher length . I guess most of the installs use a newer cipher, and nobody is using 16+ long passwords.

**Comment by alecpl** [comment]
I can't access files you linked. As for the rest I don't know. I'm using default DES-EDE3-CBC and have no problem with long passwords. Also, the cipher initialization vector length does not limit the length of a string to be encrypted. Would be nice if you could re-attach the files.

**Comment by sandormarton** [comment]
[cons2.txt](https://github.com/roundcube/roundcubemail/files/1917108/cons2.txt)
[cons1.txt](https://github.com/roundcube/roundcubemail/files/1917109/cons1.txt)
[cons1patch.txt](https://github.com/roundcube/roundcubemail/files/1917110/cons1patch.txt)

Reattached them. 
Or you can find them at : 
http://kobor.udvarhely.net/vw/rc/cons1patch.txt
http://kobor.udvarhely.net/vw/rc/cons1.txt
http://kobor.udvarhely.net/vw/rc/cons2.txt

**Comment by alecpl** [comment]
I applied your patch, faked the code to use the same ckey, method and iv as in your example, and used the same password. encrypt() result string is the same as in your example but the password is decrypted correctly. That makes me believe the problem is in PHP or OpenSSL. And you're the first person with such an issue I've heard of. I'm not sure what more we can do here.

I tested with PHP 7.0.28, OpenSSL 1.0.2g.

**Comment by sandormarton** [comment]
Ok, the problem is not Roundcube related.
I did copy out the relevant parts to a standalone app, and tested on multiple servers and multiple PHp versions; all worked except that one .
On that server i'm using Ondrej Sury's PPA for PHP. And only PHP 5.6 was showing the issue, PHP 7.x was ok. And in the end turned out that PHP 5.6 upgrades were being held back, once i upgraded to latest , the test script started to work ok.
Sorry for wasting your time.

**Comment by valerio-bozzolan** [comment]
Wow.
It would be interesting in finding the related upstream PHP bug.

**Comment by valerio-bozzolan** [comment]
Anyway, I've just found that in my case the bug seems not related to the length of the password, but it raises up using unusual character in the password, like `£` in my case.

**Comment by garry08** [comment]
If someone happens to randomly land here like I did today, I faced the same problem as @valerio-bozzolan ; a password containing a pound sign `£` not being able to authenticate.

For me the problem was that the `password_charset` was not set in the config and was defaulting to `ISO-8859-1`.

I fixed it by adding `$config['password_charset'] = 'UTF-8';` to my `config.inc.php`.

PS: This problem should slowly disappear since now `UTF-8` is the default value: https://github.com/roundcube/roundcubemail/issues/6522
